File: src/main/java/time/primitives/TimePrimitives.java
package time.primitives;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;
import org.nlogo.api.Argument;
import org.nlogo.api.Context;
import org.nlogo.api.Reporter;
import org.nlogo.api.ExtensionException;
import org.nlogo.api.LogoException;
import org.nlogo.core.Syntax;
import org.nlogo.core.SyntaxJ;
import org.nlogo.nvm.ExtensionContext;

import time.TimeExtension;
import time.TimeUtils;
import time.TimeEnums.PeriodType;
import time.datatypes.LogoTime;

public class TimePrimitives {
	public static class NewLogoTime implements Reporter {
		public Syntax getSyntax() {
			return SyntaxJ.reporterSyntax(new int[]{Syntax.StringType()},
					Syntax.WildcardType());
		}
		public Object report(Argument args[], Context context) throws ExtensionException, LogoException {
			TimeExtension.context = context; // for debugging
			LogoTime time = new LogoTime(TimeUtils.getStringFromArgument(args, 0));
			return time;
		}
	}

	public static class Anchor implements Reporter {
		public Syntax getSyntax() {
			return SyntaxJ.reporterSyntax(new int[]{Syntax.WildcardType(),Syntax.NumberType(),Syntax.StringType()},
					Syntax.WildcardType());
		}
		public Object report(Argument args[], Context context) throws ExtensionException, LogoException {
			LogoTime time = TimeUtils.getTimeFromArgument(args, 0);
			LogoTime newTime = new LogoTime(time);
			newTime.setAnchor(TimeUtils.getDoubleFromArgument(args, 1),
					TimeUtils.stringToPeriodType(TimeUtils.getStringFromArgument(args, 2)),
					((ExtensionContext)context).workspace().world());
			return newTime;
		}
	}
	
	public static class Show implements Reporter {
		public Syntax getSyntax() {
			return SyntaxJ.reporterSyntax(new int[]{Syntax.WildcardType(),Syntax.StringType()},
					Syntax.StringType());
		}
		public Object report(Argument args[], Context context) throws ExtensionException, LogoException {
			LogoTime time = TimeUtils.getTimeFromArgument(args, 0);
			String fmtString = TimeUtils.getStringFromArgument(args, 1);
			DateTimeFormatter fmt = null;
			if(fmtString.trim().equals("")){
				fmt = DateTimeFormat.forPattern("yyyy-MM-dd HH:mm:ss.SSS");
			}else{
				fmt = DateTimeFormat.forPattern(fmtString);
			}
			return time.show(fmt);
		}
	}
	
	public static class Copy implements Reporter {
		public Syntax getSyntax() {
			return SyntaxJ.reporterSyntax(new int[]{Syntax.WildcardType()},
					Syntax.WildcardType());
		}
		public Object report(Argument args[], Context context) throws ExtensionException, LogoException {
			LogoTime time = TimeUtils.getTimeFromArgument(args, 0);
			return new LogoTime(time);
		}
	}
	
	public static class CreateWithFormat implements Reporter {
		public Syntax getSyntax() {
			return SyntaxJ.reporterSyntax(new int[]{Syntax.StringType(),Syntax.StringType()},
					Syntax.WildcardType());
		}
		public Object report(Argument args[], Context context) throws ExtensionException, LogoException {
			TimeExtension.context = context; // for debugging
			LogoTime time = new LogoTime(TimeUtils.getStringFromArgument(args, 0),TimeUtils.getStringFromArgument(args, 1));
			return time;
		}
	}
	
	public static class DifferenceBetween implements Reporter {
		public Syntax getSyntax() {
			return SyntaxJ.reporterSyntax(new int[]{Syntax.WildcardType(),Syntax.WildcardType(),Syntax.StringType()},
					Syntax.NumberType());
		}
		public Object report(Argument args[], Context context) throws ExtensionException, LogoException {
			LogoTime startTime = TimeUtils.getTimeFromArgument(args,0);
			LogoTime endTime = TimeUtils.getTimeFromArgument(args,1);
			PeriodType pType = TimeUtils.stringToPeriodType(TimeUtils.getStringFromArgument(args, 2));
			return startTime.getDifferenceBetween(pType, endTime);
		}
	}
	
	public static class Get implements Reporter {
		public Syntax getSyntax() {
			return SyntaxJ.reporterSyntax(new int[]{Syntax.StringType(),Syntax.WildcardType()},
					Syntax.NumberType());
		}
		public Object report(Argument args[], Context context) throws ExtensionException, LogoException {
			PeriodType periodType = TimeUtils.stringToPeriodType(TimeUtils.getStringFromArgument(args, 0));
			LogoTime time = TimeUtils.getTimeFromArgument(args, 1);
			return time.get(periodType).doubleValue();
		}
	}
	
	public static class IsAfter implements Reporter {
		public Syntax getSyntax() {
			return SyntaxJ.reporterSyntax(new int[]{Syntax.WildcardType(),Syntax.WildcardType()},
					Syntax.BooleanType());
		}
		public Object report(Argument args[], Context context) throws ExtensionException, LogoException {
			LogoTime timeA = TimeUtils.getTimeFromArgument(args,0);
			LogoTime timeB = TimeUtils.getTimeFromArgument(args,1);
			return !(timeA.isBefore(timeB) || timeA.isEqual(timeB));
		}
	}
	
	public static class IsBefore implements Reporter {
		public Syntax getSyntax() {
			return SyntaxJ.reporterSyntax(new int[]{Syntax.WildcardType(),Syntax.WildcardType()},
					Syntax.BooleanType());
		}
		public Object report(Argument args[], Context context) throws ExtensionException, LogoException {
			LogoTime timeA = TimeUtils.getTimeFromArgument(args,0);
			LogoTime timeB = TimeUtils.getTimeFromArgument(args,1);
			return timeA.isBefore(timeB);
		}
	}
	
	public static class IsBetween implements Reporter {
		public Syntax getSyntax() {
			return SyntaxJ.reporterSyntax(new int[]{Syntax.WildcardType(),Syntax.WildcardType(),Syntax.WildcardType()},
					Syntax.BooleanType());
		}
		public Object report(Argument args[], Context context) throws ExtensionException, LogoException {
			LogoTime timeA = TimeUtils.getTimeFromArgument(args,0);
			LogoTime timeB = TimeUtils.getTimeFromArgument(args,1);
			LogoTime timeC = TimeUtils.getTimeFromArgument(args,2);
			return timeA.isBetween(timeB,timeC);
		}
	}
	
	public static class IsEqual implements Reporter {
		public Syntax getSyntax() {
			return SyntaxJ.reporterSyntax(new int[]{Syntax.WildcardType(),Syntax.WildcardType()},
					Syntax.BooleanType());
		}
		public Object report(Argument args[], Context context) throws ExtensionException, LogoException {
			LogoTime timeA = TimeUtils.getTimeFromArgument(args,0);
			LogoTime timeB = TimeUtils.getTimeFromArgument(args,1);
			return timeA.isEqual(timeB);
		}
	}
	
	public static class Plus implements Reporter {
		public Syntax getSyntax() {
			return SyntaxJ.reporterSyntax(new int[]{Syntax.WildcardType(),Syntax.NumberType(),Syntax.StringType()},
					Syntax.WildcardType());
		}
		public Object report(Argument args[], Context context) throws ExtensionException, LogoException {
			LogoTime time = new LogoTime(TimeUtils.getTimeFromArgument(args,0));
			return time.plus(TimeUtils.stringToPeriodType(TimeUtils.getStringFromArgument(args, 2)), TimeUtils.getDoubleFromArgument(args, 1));
		}
	}
}

File: src/primitives/TimePrimitives.scala
package org.nlogo.extensions.time.primitives

import java.time.format.DateTimeFormatter
import java.time.format.ResolverStyle.STRICT
import org.nlogo.api.Argument
import org.nlogo.api.Context
import org.nlogo.api.Reporter
import org.nlogo.core.Syntax
import org.nlogo.core.SyntaxJ
import org.nlogo.nvm.ExtensionContext
import org.nlogo.extensions.time._
import org.nlogo.extensions.time.datatypes._

object TimePrimitives {

  class NewLogoTime extends Reporter {
    def getSyntax(): Syntax =
      SyntaxJ.reporterSyntax(Array(Syntax.StringType), Syntax.WildcardType)
    def report(args: Array[Argument], context: Context): AnyRef = {
      TimeExtension.context = context
      new LogoTime(TimeUtils.getStringFromArgument(args, 0))
    }
  }

  class Anchor extends Reporter {
    def getSyntax(): Syntax =
      SyntaxJ.reporterSyntax(
        Array(Syntax.WildcardType, Syntax.NumberType, Syntax.StringType),
        Syntax.WildcardType)
    def report(args: Array[Argument], context: Context): AnyRef = {
      /* The anchor primitives have experienced some changes with the use of the
         Java time library. Originally, the Joda library would create copies of
         LocalDateTime, LocalDate, and MonthDay. That is no longer the case with
         the Java 8 time library, instead preferring references. This has caused
         some headaches with scheduler and timeseries functionality.
         Since two LogoTimes can be referencing the same LocalDateTime/LocalDate/
         MonthDay, this can be a source of inconsistencies.
         CBR [ 04/26/2019 ]
       */
      val newTime: LogoTime = new LogoTime(TimeUtils.getTimeFromArgument(args, 0))
      newTime.setAnchor(
        TimeUtils.getDoubleFromArgument(args, 1),
        TimeUtils.stringToPeriodType(TimeUtils.getStringFromArgument(args, 2)),
        context.asInstanceOf[ExtensionContext].workspace.world
      )
      newTime
    }
  }

  class Show extends Reporter {
    def getSyntax(): Syntax =
      SyntaxJ.reporterSyntax(Array(Syntax.WildcardType, Syntax.StringType),
                             Syntax.StringType)
    def report(args: Array[Argument], context: Context): AnyRef = {
      val time: LogoTime = TimeUtils.getTimeFromArgument(args, 0)
      /* There is an implicit LogoTime initialization */
      val fmtString: String = TimeUtils.getStringFromArgument(args, 1)
      val fmt =
        if (fmtString.trim().==(""))
          DateTimeFormatter.ofPattern("uuuu-MM-dd HH:mm:ss.SSS").withResolverStyle(STRICT)
        else DateTimeFormatter.ofPattern(fmtString.replace('y','u')).withResolverStyle(STRICT)
      time.show(fmt)
    }
  }

  class Copy extends Reporter {
    def getSyntax(): Syntax =
      SyntaxJ.reporterSyntax(Array(Syntax.WildcardType),
                             Syntax.WildcardType)
    def report(args: Array[Argument], context: Context): AnyRef = {
      val time: LogoTime = TimeUtils.getTimeFromArgument(args, 0)
      new LogoTime(time)
    }
  }

  class CreateWithFormat extends Reporter {
    def getSyntax(): Syntax =
      SyntaxJ.reporterSyntax(Array(Syntax.StringType, Syntax.StringType),
                             Syntax.WildcardType)
    def report(args: Array[Argument], context: Context): AnyRef = {
      TimeExtension.context = context
      val customFormat = TimeUtils.getStringFromArgument(args, 1) match {
        case null => None
        case x => Some(x)
      }
      new LogoTime(TimeUtils.getStringFromArgument(args, 0),customFormat)
    }
  }

  class DifferenceBetween extends Reporter {
    def getSyntax(): Syntax =
      SyntaxJ.reporterSyntax(Array(Syntax.WildcardType,
                                   Syntax.WildcardType,
                                   Syntax.StringType),
                                   Syntax.NumberType)
    def report(args: Array[Argument], context: Context): AnyRef = {
      val startTime: LogoTime = TimeUtils.getTimeFromArgument(args, 0)
      val endTime: LogoTime = TimeUtils.getTimeFromArgument(args, 1)
      val pType: PeriodType =
        TimeUtils.stringToPeriodType(TimeUtils.getStringFromArgument(args, 2))
      startTime.getDifferenceBetween(pType, endTime)
    }
  }

  class Get extends Reporter {
    def getSyntax(): Syntax =
      SyntaxJ.reporterSyntax(Array(Syntax.StringType, Syntax.WildcardType), Syntax.NumberType)
    def report(args: Array[Argument], context: Context): AnyRef = {
      val periodType: PeriodType =
        TimeUtils.stringToPeriodType(TimeUtils.getStringFromArgument(args, 0))
      val time: LogoTime = TimeUtils.getTimeFromArgument(args, 1)
      time.get(periodType).doubleValue().asInstanceOf[AnyRef]
    }
  }

  class IsAfter extends Reporter {
    def getSyntax(): Syntax =
      SyntaxJ.reporterSyntax(
        Array(Syntax.WildcardType, Syntax.WildcardType),
        Syntax.BooleanType)
    def report(args: Array[Argument], context: Context): AnyRef = {
      val timeA: LogoTime = TimeUtils.getTimeFromArgument(args, 0)
      val timeB: LogoTime = TimeUtils.getTimeFromArgument(args, 1)
      (!(timeA.isBefore(timeB) || timeA.isEqual(timeB))).asInstanceOf[AnyRef]
    }
  }

  class IsBefore extends Reporter {
    def getSyntax(): Syntax =
      SyntaxJ.reporterSyntax(
        Array(Syntax.WildcardType, Syntax.WildcardType),
        Syntax.BooleanType)
    def report(args: Array[Argument], context: Context): AnyRef = {
      val timeA: LogoTime = TimeUtils.getTimeFromArgument(args, 0)
      val timeB: LogoTime = TimeUtils.getTimeFromArgument(args, 1)
      timeA.isBefore(timeB).asInstanceOf[AnyRef]
    }
  }

  class IsBetween extends Reporter {
    def getSyntax(): Syntax =
      SyntaxJ.reporterSyntax(Array(Syntax.WildcardType,
                                   Syntax.WildcardType,
                                   Syntax.WildcardType),
                                   Syntax.BooleanType)
    def report(args: Array[Argument], context: Context): AnyRef = {
      val timeA: LogoTime = TimeUtils.getTimeFromArgument(args, 0)
      val timeB: LogoTime = TimeUtils.getTimeFromArgument(args, 1)
      val timeC: LogoTime = TimeUtils.getTimeFromArgument(args, 2)
      timeA.isBetween(timeB, timeC).asInstanceOf[AnyRef]
    }
  }

  class IsEqual extends Reporter {
    def getSyntax(): Syntax =
      SyntaxJ.reporterSyntax(
        Array(Syntax.WildcardType, Syntax.WildcardType),
        Syntax.BooleanType)
    def report(args: Array[Argument], context: Context): AnyRef = {
      val timeA: LogoTime = TimeUtils.getTimeFromArgument(args, 0)
      val timeB: LogoTime = TimeUtils.getTimeFromArgument(args, 1)
      timeA.isEqual(timeB).asInstanceOf[AnyRef]
    }
  }

  class Plus extends Reporter {
    def getSyntax(): Syntax =
      SyntaxJ.reporterSyntax(
        Array(Syntax.WildcardType, Syntax.NumberType, Syntax.StringType),
        Syntax.WildcardType)
    def report(args: Array[Argument], context: Context): AnyRef = {
      val time: LogoTime = new LogoTime(TimeUtils.getTimeFromArgument(args, 0))
      time.plus(
        TimeUtils.stringToPeriodType(TimeUtils.getStringFromArgument(args, 2)),
        TimeUtils.getDoubleFromArgument(args, 1))
    }
  }
}
